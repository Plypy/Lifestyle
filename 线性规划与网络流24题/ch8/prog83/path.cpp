/* 可以转化为二分图匹配模型
 * 因为可以这样考虑，对于图中的任意一个顶点
 * 若其不为某一个路径的终点的话，那么从这个点必然有出边
 * 也就是说必然有另一个顶点与这个顶点相对应
 * 这也就提示了我们二分图最大匹配的模型可能适用
 * 但是题目要求的是最小路径覆盖数,这个怎么考虑呢
 * 考虑到每条路径必然会有一个终点没有匹配上
 * 也就是说没匹配上的顶点个数也就是路径数
 * 那么要求最小路径覆盖的话，只需要让没匹配上的顶点最少，即匹配的最多即可
 * 最后答案就是顶点数减去最大匹配数
 */
#include <iostream>
#include <cstdlib>
#include <cstdio>
#include <fstream>
using namespace std;
ifstream fin("path.in");
ofstream fout("path.out");
const int MAXN = 150;

bool g[MAXN][MAXN];
bool used[MAXN];
int pal[MAXN];
int nxt[MAXN];
int n, m;

void load() {
    fin >> n >> m;
    while (m--) {
        int a, b;
        fin >> a >> b;
        g[a][b] = true;
    }
}
int ans;
bool find(int u) {
    for (int v = 1; v <= n; ++v)
        if (g[u][v] && !used[v]) {
            used[v] = true;
            if (!pal[v] || find(pal[v])) {
                nxt[u] = v;
                pal[v] = u;
                return true;
            }
        }
    return false;
}
void work() {
    for (int i = 1; i <= n; ++i) {
        memset(used,0,sizeof(used));
        if (find(i))
            ++ans;
    }
    ans = n-ans;
}

void output() {
    for (int i = 1; i <= n; ++i) {
        if (!pal[i]) {
            int t = i;
            while (true) {
                fout << t;
                if (!nxt[t]) {
                    fout.put('\n');
                    break;
                }
                t = nxt[t];
                fout.put(' ');
            }
        }
    }
    fout << ans << endl;
}

int main() {
    load();
    work();
    output();
    return 0;
} 
