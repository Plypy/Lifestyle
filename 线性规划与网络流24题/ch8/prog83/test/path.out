1 3 6 43 78 95 109
2 4 13 20 23 25 33 45 62 84 94
5 14 18 38 39 58 69 107
7 8 12 16 24 64 65 68 73 75 86 87 92 96 108
9 41 47 63 67 74 89 100
10 17 29 37 46 51 54 76 99 101
11 19 31 40 53 71 83 93 104
15 26 30 56 57 60 70 90 103
21 22 35 36 49 59 81 88 106
27 52 66 72 80 85 111
28 42 50 77 98 110
32 34 61 79 91 97 105
44 48 55 82 102
13
